Shorten parameter lists by containing StorageEngine, TransactionIdStore, and LogFiles into new CoreBuildStore class

rabex028 opened the pull request:
This request addresses issue #1.

The purpose of the changes in this request is to mitigate a code design flaw, or *smell*, known as long parameter lists in the `NeoStoreDataSource` class. Long parameter lists occur when an excessive number of arguments are passed into a class or method, which can result in code that is both harder to read and maintain. 

We are focusing on `NeoStoreDataSource.java` for this refactoring because it was identified by an online tool called codescene as being a refactoring target. Codescene determined that the code in this class is relatively complicated, and also that changes are frequently made to it. Because of these two factors, a significant amount of development time and effort likely goes into working on `NeoStoreDataSource.java`. Refactoring this class to make it easier to understand can therefore reduce the amount of development resources that are spent on this code over the long run.

The long parameter list smell is being addressed because an IDE tool designed to search for code smells, named Sonar Lint, flagged several cases of it. Sonar Lint recommends that a maximum of 7 parameters are given to methods or classes, and there are multiple parameter lists which exceed this amount in `NeoStoreDataSource.java`.

Our general approach to address this smell has been to find groups of parameters that are common within multiple parameter lists, enclose them within a new class, and pass this class in as a parameter instead.

* Commit 74106c8 creates a new class we have named `CoreBuildStore` to contain the parameters `StorageEngine`, `TransactionIdStore`, and `LogFiles`. These three parameters are then removed from the list given to the method `buildRecovery()` and are replaced with `CoreBuildStore`.

* Commit d9e1d33 performs a similar refactoring by replacing `StorageEngine`, `TransactionIdStore`, and `LogFiles` again with `CoreBuildStore` in the `buildKernel()` and `buildTransactionLogs()` methods' parameter lists. In addition, unnecessary redundant variable assignments and deprecated method calls were removed as recommended by Sonar Lint.

nickplucker committed:
Create separate class to group common objects related to buildRecovery() method

"StorageEngine", "TransactionIdStore" and "LogFiles" were moved to the new class (CoreBuildStore) as they are common in three methods (buildRecovery(), buildKernel() and buildTransactionLogs()). This specific commit addresses the long parameter list in buildRecovery() and CoreBuildStore will be used to address long parameter lists in the other two methods in subsequent commits. All tests pass.

nickplucker committed:
Merge remote-tracking branch 'remotes/Neo4j-master/3.5' into LongParameterListSmell

nickplucker committed:
Refactor buildKernel() and buildTransactionLogs() by moving StorageEngine, TransactionIdStore and LogFiles to CoreBuildStore

We also cleaned up unnecessary items as suggested by SonarLint, such as redundant variable assignments and deprecated methods.

YujingSong reviewed (approved):
It is a good idea to create a new class in your case.  Looks like did some reorganization in the original main class, and combined some functions together.  Very good job.

schon094 reviewed (approved):
Creating a new class to contain common objects is a good way to reduce the amount of parameters passed to these methods.
Nice job cleaning up unnecessary variables as well.

CMenne commented on community/kernel/src/main/java/org/neo4j/kernel/CoreBuildStore.java:
```diff
@@ -0,0 +1,42 @@
+import org.neo4j.storageengine.api.StorageEngine;
+import org.neo4j.kernel.impl.transaction.log.TransactionIdStore;
+
+public class CoreBuildStore {
```
In it's current iteration, this class is a 'data class'. Potentially in the future, logic could be migrated into the class and if not, it might be more valuable to revert these changes.

CMenne commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -344,7 +345,7 @@ public void start() throws IOException
 
         // Build all modules and their services
-        StorageEngine storageEngine = null;
+        storageEngine = null;
```
:+1:

CMenne commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -530,22 +532,22 @@ private NeoStoreTransactionLogModule buildTransactionLogs( LogFiles logFiles, Co
 
         final LogPruning logPruning =
-                new LogPruningImpl( fs, logFiles, logProvider, new LogPruneStrategyFactory(), clock, config );
+                new LogPruningImpl( fs, coreBuildStore.getLogFiles(), logProvider, new LogPruneStrategyFactory(), clock, config );
```
While you reduced message chaining in other parts of the code, you've added it to numerous other parts of the code.

CMenne commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -518,10 +520,10 @@ private StorageEngine buildStorageEngine(
 
-    private NeoStoreTransactionLogModule buildTransactionLogs( LogFiles logFiles, Config config,
-            LogProvider logProvider, JobScheduler scheduler, StorageEngine storageEngine,
+    private NeoStoreTransactionLogModule buildTransactionLogs( CoreBuildStore coreBuildStore, Config config,
```
By reducing the number of parameters, the code is easier to read.

CMenne reviewed (changes requested):
While we feel that the changes you made reduced the number of parameters in certain methods and thus improved readability, we feel that this wasn't a beneficial change due to the fact that message chains were introduced into other areas. On it's own these changes are not sufficiently beneficial to warrant accepting the pull request. If logic was integrated into CoreBuildStore, this would remedy the situation. 

@Xaivteev @CMenne @twarling

bemica reviewed (approved):
I agree that CoreBuildStore is a data class, and it would be nice to see some functionality in it (which would also help to reduce message chains) but this is still a good improvement over the previous build, and is a good stepping stone for further refactoring

stev0531 reviewed (approved):
I also agree that `CoreBuildStore` is a data class right now, but as was also already mentioned, it's a good stepping stone to refactor further. The pull request message, though, was really well done. You explain why you chose the smell you did, how you planned to address it, the tools you used to find it, and even what the tools were. I found this really nice because if I didn't have any previous knowledge about the tools or what you planned to do, I still would have a good understanding of your reasoning for why your changes matter.

rabex028 committed:
Move transactionIdStore method into CoreBuildStore

A method which initializes the versionContextSupplier with
the last closed transactionId has been added to CoreBuildStore
to help prevent it from being a data class. Addresses feedback from
PR #2

globbycodes commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -359,45 +360,46 @@ public void start() throws IOException
                     versionContextSupplier );
             life.add( logFiles );
 
+            CoreBuildStore coreBuildStore = new CoreBuildStore();
```
I think this is a nice idea of reducing the list of parameters by creating `CoreBuildStore` puttin everything into that class.
I don't know what `logFiles`. `storageEngine` and `transactionIdStore` really are. But I'm assuming that these are related and it makes sense to put them together.

globbycodes commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -530,22 +532,22 @@ private NeoStoreTransactionLogModule buildTransactionLogs( LogFiles logFiles, Co
 
         final LogPruning logPruning =
-                new LogPruningImpl( fs, logFiles, logProvider, new LogPruneStrategyFactory(), clock, config );
+                new LogPruningImpl( fs, coreBuildStore.getLogFiles(), logProvider, new LogPruneStrategyFactory(), clock, config );
```
As an option you could have done something like that:
```Java
LogFiles logFiles = coreBuildStore.getLogFiles()
```
and then use that instead of calling  `coreBuildStore.getLogFiles()` every time.

CMenne replied:
👍

globbycodes reviewed (approved):
I think overall you have done a good job. The list of parameters as mentioned actually got reduced and therefore made the code a bit easier to understand. The documentation of the pull request is good. I could understand the big picture of what was the problem.

rabex028 committed:
Fix parenthesis spacing in CoreBuildStore class

nickplucker commented:
@CMenne Would you be able to review our PR again? We implemented some changes that you requested.

CMenne reviewed (approved):
After further review, this does seem like an improvement. I agree, this does reduce the total number of parameters in these methods.
